### PR TITLE
Debug CI tests on Ada

### PR DESCRIPTION
**Pull request by timmoon10**
This applies the changes in https://github.com/NVIDIA/TransformerEngine/pull/393 to the PyTorch and Paddle tests. In particular, it only runs tests involving cuDNN fused attention on compute capabilities 8.0 and 9.0.

### CODE REVIEW

**Comment by timmoon10** [comment]
Pipeline 9489089

**Comment by timmoon10** [comment]
Running on an L40, I found that the JAX FP8 GEMM tests on integer matrices were failing. It seems cuBLAS chooses a split-k kernel that prevents us from getting bit-wise correct results, although it is still within the expected FP8 error. I've changed the matrix dims to help cuBLAS pick a nicer kernel.

Pipeline 9504876.

**Comment by timmoon10** [comment]
Pipeline 9617488 is green.

File: tests/pytorch/test_fused_attn.py
Comment (cyanguwa) [inline tests/pytorch/test_fused_attn.py:417]: Unfinished sentence here?

File: tests/jax/test_custom_call_compute.py
Comment (cyanguwa) [inline tests/jax/test_custom_call_compute.py:30]: Is there a reason to make these matrices smaller?

File: tests/jax/test_fused_attn.py
Comment (cyanguwa) [inline tests/jax/test_fused_attn.py:151]: Hey, I think there might be some misunderstanding about the supported architecture for fused attention: https://github.com/NVIDIA/TransformerEngine/pull/393#discussion_r1300617951

The max512 backend doesn't support sm86/89 but the arbitrary_seqlen backend does, since cuDNN 8.9.3.

I think you can use something like this to check whether the compute capability condition is met. (This should work for PyTorch at least, Jax might have its own extension with a similar name.)

https://github.com/NVIDIA/TransformerEngine/blob/a150d286a76f6cd4e34b0e16e8b9cc11c4dba34c/transformer_engine/pytorch/attention.py#L1150

**Review by cyanguwa — CHANGES_REQUESTED** [review]

File: tests/jax/test_custom_call_compute.py
Comment (timmoon10) [reply tests/jax/test_custom_call_compute.py:30]: https://github.com/NVIDIA/TransformerEngine/pull/397#issuecomment-1692492335

The tests with integer matrices failed with the larger sizes when I ran on an L40. cuBLAS picked GEMM kernels that were not quite accurate enough to get bit-wise exact results, I think because it did split-k GEMM. I just reduced the matrix sizes to get back to bit-wise exact results, but we could alternatively modify the integer matrix tests with FP8 numerical tolerances.

File: tests/jax/test_fused_attn.py
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:151]: Using `get_fused_attn_backend` for checking if fused attention supportness has been updated in #393, and it has been merged into the main branch

File: tests/pytorch/test_fused_attn.py
Comment (timmoon10) [reply tests/pytorch/test_fused_attn.py]: 🤦

**Comment by timmoon10** [comment]
I've tweaked the PyTorch and JAX fused attention tests so we check if there's a supported backed (namely `F16_arbitrary_seqlen` on Ada). These pass when I run manually on an L40 and I've launched pipeline 9938409.

https://github.com/NVIDIA/TransformerEngine/pull/403 adds some PyTorch attention tests and https://github.com/NVIDIA/TransformerEngine/pull/411 adds backend detection logic to Paddle. We should hold off on merging until those are in.

**Comment by timmoon10** [comment]
This PR is now good to go, pending pipeline 10094388.

File: tests/pytorch/test_fused_attn.py
Comment (cyanguwa) [inline tests/pytorch/test_fused_attn.py:544]: I think these `if backend ==` statements should be kept here, because it's in the `_run_dpa_fp8_ref` function that we call DPA, and the `_run_dpa_fp8` calls something custom (`DPA_FP8()` defined in this test).

**Review by cyanguwa — APPROVED** [review]
Just that one comment, otherwise looks good!

File: tests/paddle/test_operators.py
Comment (timmoon10) [inline tests/paddle/test_operators.py:768]: These tolerances match the PyTorch tests:
https://github.com/NVIDIA/TransformerEngine/blob/daa5e1840571be4f9cd11053bbde3f1bfd8ca489/tests/pytorch/test_fused_attn.py#L86

I suspect there might be a bug in how cuDNN fused attention handles causal masks. With the old tolerances, the test passes on H100 (max-512 seqlen backend) and fails on L40 (arbitrary backend). I tried changing the attention mask so that it is a plain triangular matrix that does not take the actual sequence lengths into account, i.e. modified 
https://github.com/NVIDIA/TransformerEngine/blob/777e45be0b36206557c76280451bf94942637115/tests/paddle/test_operators.py#L605
to loop over `range(self.q_seqlen)`. This resulted in the test passing on L40 and failing on H100. This makes me think that the two backends are subtly inconsistent. I'll continue digging in, but if the fix is difficult then I think we should consolidate these changes and incorporate the fix in a future PR.

Pinging @cyanguwa.

Comment (timmoon10) [reply tests/paddle/test_operators.py]: It turns out this is because the artibrary seqlen backend assumes that all seqlens are equal. However, this inconsistency is handled in https://github.com/NVIDIA/TransformerEngine/pull/403, which has more expressive layout configurations. Attempting to run with a variable seqlen format (e.g. `TH3D`) on a system that only supports the arbitrary seqlen backend will result in a runtime error. Upgrading Paddle to use the new attention layout formats is a higher priority than fixing this edge case.

**Comment by timmoon10** [comment]
Tests passed in pipeline 10211932.

**Review by ksivaman — APPROVED** [review]
Looks good

**Pull request merged**
